### PR TITLE
Added DIY callback plugin

### PR DESCRIPTION
**Pull request by theque5t**
##### SUMMARY
Customize the output via templating
Fixes: ansible/ansible#14367
Fixes: ansible/ansible#32584

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
DIY Callback Plugin

##### ADDITIONAL INFORMATION
Originally submitted here ansible/ansible#65844

Tested on the following Python versions:
- 2.7.15rc1
- 3.5.8
- 3.6.7

Tested on the following Ansible versions:
- 2.10.0.dev0
- 2.9.2
- 2.8.5

Example use case: [ansible-playbook-execute](https://github.com/theque5t/ansible-playbook-execute)

Allows the user to fully customize the output using templates supplied via Ansible configuration, environment variables, and Ansible variables(extra vars, play vars, tasks vars etc.). I've copied the example from the plugin documentation down below.

__ansible.cfg__
```
  # Enable plugin
  [defaults]
  stdout_callback=diy

  [callback_diy]
  # Output when playbook starts
  playbook_on_start_msg="diy output(via ansible.cfg): playbook example: {{ ansible_callback_diy.playbook.file_name }}"
  playbook_on_start_msg_color=yellow

  # Comment out to allow default plugin output
  # playbook_on_play_start_msg="PLAY: starting play {{ ansible_callback_diy.play.name }}"

  # Accept on_skipped_msg or ansible_callback_diy_runner_on_skipped_msg as input vars
  # If neither are supplied, omit the option
  runner_on_skipped_msg="{{ on_skipped_msg | default(ansible_callback_diy_runner_on_skipped_msg) | default(omit) }}"

  # Newline after every callback
  # on_any_msg='{{ " " | join("\n") }}'
```
__playbook.yml__
```yaml
  ---
  - name: "default plugin output: play example"
    hosts: localhost
    gather_facts: no
    tasks:
      - name:  default plugin output
        debug:
          msg: default plugin output

  - name: override from play vars
    hosts: localhost
    gather_facts: no
    vars:
      green: "\e[0m\e[38;5;82m"
      yellow: "\e[0m\e[38;5;11m"
      bright_purple: "\e[0m\e[38;5;105m"
      cyan: "\e[0m\e[38;5;51m"
      green_bg_black_fg: "\e[0m\e[48;5;40m\e[38;5;232m"
      yellow_bg_black_fg: "\e[0m\e[48;5;226m\e[38;5;232m"
      purple_bg_white_fg: "\e[0m\e[48;5;57m\e[38;5;255m"
      cyan_bg_black_fg: "\e[0m\e[48;5;87m\e[38;5;232m"
      magenta: "\e[38;5;198m"
      white: "\e[0m\e[38;5;255m"
      ansible_callback_diy_playbook_on_play_start_msg: "\n{{green}}diy output(via play vars): play example: {{magenta}}{{ansible_callback_diy.play.name}}\n\n"
      ansible_callback_diy_playbook_on_task_start_msg: "diy output(via play vars): task example: {{ ansible_callback_diy.task.name }}"
      ansible_callback_diy_playbook_on_task_start_msg_color: cyan
      ansible_callback_diy_playbook_on_stats_msg: |+2
                CUSTOM STATS
        ==============================
        {% for key in ansible_callback_diy.stats | sort %}
        {% if ansible_callback_diy.stats[key] %}
        {% if key == 'ok' %}
        {% set color_one = lookup('vars','green_bg_black_fg') %}
        {% set prefix = '      ' %}
        {% set suffix = '     ' %}
        {% set color_two = lookup('vars','green') %}
        {% elif key == 'changed' %}
        {% set color_one = lookup('vars','yellow_bg_black_fg') %}
        {% set prefix = '   ' %}
        {% set suffix = '   ' %}
        {% set color_two = lookup('vars','yellow') %}
        {% elif key == 'processed' %}
        {% set color_one = lookup('vars','purple_bg_white_fg') %}
        {% set prefix = '  ' %}
        {% set suffix = '  ' %}
        {% set color_two = lookup('vars','bright_purple') %}
        {% elif key == 'skipped' %}
        {% set color_one = lookup('vars','cyan_bg_black_fg') %}
        {% set prefix = '   ' %}
        {% set suffix = '   ' %}
        {% set color_two = lookup('vars','cyan') %}
        {% else %}
        {% set color_one = "" %}
        {% set prefix = "" %}
        {% set suffix = "" %}
        {% set color_two = "" %}
        {% endif %}
        {{ color_one }}{{ "%s%s%s" | format(prefix,key,suffix) }}{{ color_two }}: {{ ansible_callback_diy.stats[key] | to_nice_yaml }}
        {% endif %}
        {% endfor %}

    tasks:
      - name: custom banner with default plugin result output
        debug:
          msg: "default plugin output: result example"

      - name: override from task vars
        debug:
          msg: "example {{ two }}"
        changed_when: true
        vars:
          white_fg_red_bg: "\e[0m\e[48;5;1m"
          two: "{{ white_fg_red_bg }}    2    "
          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
          ansible_callback_diy_playbook_on_task_start_msg_color: bright magenta
          ansible_callback_diy_runner_on_ok_msg: "diy output(via task vars): result example: \n{{ ansible_callback_diy.result.output.msg }}\n"
          ansible_callback_diy_runner_on_ok_msg_color: "{{ 'yellow' if ansible_callback_diy.result.is_changed else 'bright green' }}"

      - name: suppress output
        debug:
          msg: i should not be displayed
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: ""
          ansible_callback_diy_runner_on_ok_msg: ""

      - name: using alias vars (see ansible.cfg)
        debug:
          msg:
        when: False
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: ""
          on_skipped_msg: "diy output(via task vars): skipped example:\n\e[0m\e[38;5;4m\u25b6\u25b6 {{ ansible_callback_diy.result.task.name }}\n"
          on_skipped_msg_color: white

      - name: just stdout
        command: echo some stdout
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: "\n"
          ansible_callback_diy_runner_on_ok_msg: "{{ ansible_callback_diy.result.output.stdout }}\n"

      - name: multline output
        debug:
          msg: "{{ multiline }}"
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
          multiline: "line\nline\nline"
          ansible_callback_diy_runner_on_ok_msg: |+2
            some
            {{ ansible_callback_diy.result.output.msg }}
            output

          ansible_callback_diy_playbook_on_task_start_msg_color: bright blue

      - name: indentation
        debug:
          msg: "{{ item.msg }}"
        with_items:
          - { indent: 1, msg: one., color: red }
          - { indent: 2, msg: two.., color: yellow }
          - { indent: 3, msg: three..., color: bright yellow }
        vars:
          ansible_callback_diy_runner_item_on_ok_msg: "{{ ansible_callback_diy.result.output.msg | indent(item.indent, True) }}"
          ansible_callback_diy_runner_item_on_ok_msg_color: "{{ item.color }}"
          ansible_callback_diy_runner_on_ok_msg: "GO!!!"
          ansible_callback_diy_runner_on_ok_msg_color: bright green

      - name: using lookup and template as file
        shell: "echo {% raw %}'output from {{ file_name }}'{% endraw %} > {{ file_name }}"
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
          file_name: diy_file_template_example
          ansible_callback_diy_runner_on_ok_msg: "{{ lookup('template', file_name) }}"

      - name: 'look at top level vars available to the "runner_on_ok" callback'
        debug:
          msg: ''
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
          ansible_callback_diy_runner_on_ok_msg: |+2
            {% for var in (ansible_callback_diy.top_level_var_names|reject('match','vars|ansible_callback_diy.*')) | sort %}
            {{ green }}{{ var }}:
              {{ white }}{{ lookup('vars', var) }}

            {% endfor %}
          ansible_callback_diy_runner_on_ok_msg_color: white

      - name: 'look at event data available to the "runner_on_ok" callback'
        debug:
          msg: ''
        vars:
          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
          ansible_callback_diy_runner_on_ok_msg: |+2
            {% for key in ansible_callback_diy | sort %}
            {{ green }}{{ key }}:
              {{ white }}{{ ansible_callback_diy[key] }}

            {% endfor %}
```

### CODE REVIEW

**Comment by theque5t** [comment]
Cc: @felixfontein @Andersson007 @gundalow

**Re: Shippable results**
I think there may be a bug in the test that failed (https://app.shippable.com/github/ansible-collections/community.general/runs/1755/2/tests) because the result mentions 134 failures associated to other components (e.g. `plugins/modules/cloud/digital_ocean/digital_ocean.py:0:0: deprecation-mismatch: "meta/runtime.yml" and DOCUMENTATION.deprecation do not agree on removal date: 'TBD' vs. None`)

**Re: ansible/third-party-check results**
I think there is a known bug in the test ( `Confirm collection can be imported into galaxy` ) that failed

ready_for_review

**Comment by felixfontein** [comment]
@theque5t you managed to create your PR in the short time between ansible/ansible#69680 was merged and #431 was merged to update tests/sanity/ignore-2.10.txt :) I'll restart the Shippable tests.

The third-party-check is currently broken, you can simply ignore it (#402).

File: plugins/callback/diy.py [hunk 1]
```diff
@@ -0,0 +1,1421 @@
...
+  callback_type: stdout
+  short_description: Customize the output
+  description:
+    - Callback plugin for the DIYer, because it allows you to supply your own custom callback templates to be output
+  version_added: "2.10"
+  author: Trevor Highfill (@theque5t)
+  extends_documentation_fragment:
+    - default_callback
+  notes:
+    - Uses the L(default,./default.html) callback plugin output when a custom callback message(C(msg)) is not provided.
```
Comment (felixfontein) [inline plugins/callback/diy.py:24]: That probably won't work anymore.
```suggestion
    - Uses the C(default) callback plugin output when a custom callback message(C(msg)) is not provided.
```

File: plugins/callback/diy.py [hunk 2]
```diff
@@ -0,0 +1,1421 @@
...
+    - Makes the callback event data available via the C(ansible_callback_diy) dictionary, which can be used in the templating context for the options.
```
Comment (felixfontein) [inline plugins/callback/diy.py:25]: Option names are typeset with `I(...)`, values with `C(...)`.
```suggestion
    - Makes the callback event data available via the I(ansible_callback_diy) dictionary, which can be used in the templating context for the options.
```

File: plugins/callback/diy.py [hunk 3]
```diff
@@ -0,0 +1,1421 @@
...
+    - For each respective callbacks C(msg) option, the following conditions result in the following the effects. If the option value C(is None or omit)
+      then the option is not being used, and therefore use the L(default,./default.html) callback plugin. If the option value C(is not None and not omit and
+      length is not greater than 0) then the option is being used without output, and therefore suppress output. If the option value C(is not None and not omit
+      and length is greater than 0) then the option is being used with output, and therefore output.
+    - "Valid color values: C(black), C(bright gray), C(blue), C(white), C(green), C(bright blue), C(cyan), C(bright green), C(red), C(bright cyan),
+      C(purple), C(bright red), C(yellow), C(bright purple), C(dark gray), C(bright yellow), C(magenta), C(bright magenta), C(normal)"
+  seealso:
+    - name: default – default Ansible screen output
+      description: The official documentation on the B(default) callback plugin.
+      link: ./default.html
```
Comment (felixfontein) [inline plugins/callback/diy.py:43]: It's probably best to link to https://docs.ansible.com/ansible/latest/plugins/callback/default.html for now.

File: plugins/callback/diy.py [hunk 4]
```diff
@@ -0,0 +1,1421 @@
...
+        - name: ansible_callback_diy_playbook_on_setup_msg_color
+      required: False
+      type: string
+'''
+
+EXAMPLES = r'''
+ansible.cfg: >
+  # Enable plugin
+  [defaults]
+  stdout_callback=diy
```
Comment (felixfontein) [inline plugins/callback/diy.py:650]:
```suggestion
  stdout_callback=community.general.diy
```

File: plugins/callback/diy.py [hunk 5]
```diff
@@ -0,0 +1,1421 @@
...
+    """
+    Callback plugin for the DIYer, because it allows you to supply your own custom callback templates to be output
+    """
+    CALLBACK_VERSION = 2.0
+    CALLBACK_TYPE = 'stdout'
+    CALLBACK_NAME = 'diy'
+
+    DIY_NS = 'ansible_callback_diy'
+
+    class DummyStdout(object):
```
Comment (felixfontein) [inline plugins/callback/diy.py:849]: Why not move this one level up?

File: plugins/callback/diy.py [hunk 6]
```diff
@@ -0,0 +1,1421 @@
...
+      respective callback.
+      Use the C(ansible_callback_diy) dictionary to see what is available to a callback. Additionally, C(ansible_callback_diy.top_level_var_names) will output
+      the top level variable names available to the callback.
+    - Each option value is rendered as a template before being evaluated. This allows for the dynamic usage of an option. For example,
+      C("{{ 'yellow' if ansible_callback_diy.result.is_changed else 'bright green' }}")
+    - For each respective callbacks C(msg) option, the following conditions result in the following the effects. If the option value C(is None or omit)
+      then the option is not being used, and therefore use the L(default,./default.html) callback plugin. If the option value C(is not None and not omit and
+      length is not greater than 0) then the option is being used without output, and therefore suppress output. If the option value C(is not None and not omit
+      and length is greater than 0) then the option is being used with output, and therefore output.
```
Comment (Andersson007) [inline plugins/callback/diy.py:37]:
```suggestion
    - For each respective callbacks C(msg) option, the following conditions result in the following the effects.
      If the option value C(is None or omit), then the option is not being used, and,
      therefore, use the L(default,./default.html) callback plugin.
      If the option value C(is not None and not omit and length is not greater than 0),
      then the option is being used without output, and therefore suppress output.
      If the option value C(is not None and not omit and length is greater than 0),
      then the option is being used with output, and therefore output.
```
Would be cool if there aren't such bunches of text written as one long line: difficult to read and focus on their meaning. I suggest refactoring of this section.
I'd also add commas to `if ..., then ...` statements.

File: plugins/callback/diy.py [hunk 7]
```diff
@@ -0,0 +1,1421 @@
...
+      C(purple), C(bright red), C(yellow), C(bright purple), C(dark gray), C(bright yellow), C(magenta), C(bright magenta), C(normal)"
+  seealso:
+    - name: default – default Ansible screen output
+      description: The official documentation on the B(default) callback plugin.
+      link: ./default.html
+  requirements:
+    - set as stdout_callback in configuration
+  options:
+    on_any_msg:
+      description: Output to be used for callback on_any
```
Comment (Andersson007) [inline plugins/callback/diy.py:48]:
```suggestion
      description: Output to be used for callback on_any.
```

File: plugins/callback/diy.py [hunk 8]
```diff
@@ -0,0 +1,1421 @@
...
+        - name: ANSIBLE_CALLBACK_DIY_ON_ANY_MSG
+      vars:
+        - name: ansible_callback_diy_on_any_msg
+      required: False
+      type: string
+
+    on_any_msg_color:
+      description:
+        - Output color to be used for on_any_msg
+        - Template should render a L(valid color value,#notes)
```
Comment (Andersson007) [inline plugins/callback/diy.py:62]:
```suggestion
        - Output color to be used for on_any_msg.
        - Template should render a L(valid color value,#notes).
```
the same for the other descriptions

File: plugins/callback/diy.py [hunk 9]
```diff
@@ -0,0 +1,1421 @@
...
+      description: Output to be used for callback on_any
+      ini:
+        - section: callback_diy
+          key: on_any_msg
+      env:
+        - name: ANSIBLE_CALLBACK_DIY_ON_ANY_MSG
+      vars:
+        - name: ansible_callback_diy_on_any_msg
+      required: False
+      type: string
```
Comment (Andersson007) [inline plugins/callback/diy.py:57]:
```suggestion
      type: str
```
`required: no` is the default AFAIK

File: plugins/callback/diy.py [hunk 10]
```diff
@@ -0,0 +1,1421 @@
...
+        - Template should render a L(valid color value,#notes)
+      ini:
+        - section: callback_diy
+          key: on_any_msg_color
+      env:
+        - name: ANSIBLE_CALLBACK_DIY_ON_ANY_MSG_COLOR
+      vars:
+        - name: ansible_callback_diy_on_any_msg_color
+      required: False
+      type: string
```
Comment (Andersson007) [inline plugins/callback/diy.py:71]:
```suggestion
      type: str
```
the same for the other parameters

File: plugins/callback/diy.py [hunk 11]
```diff
@@ -0,0 +1,1421 @@
...
+  # Accept on_skipped_msg or ansible_callback_diy_runner_on_skipped_msg as input vars
+  # If neither are supplied, omit the option
+  runner_on_skipped_msg="{{ on_skipped_msg | default(ansible_callback_diy_runner_on_skipped_msg) | default(omit) }}"
+
+  # Newline after every callback
+  # on_any_msg='{{ " " | join("\n") }}'
+
+playbook.yml: >
+  ---
+  - name: "default plugin output: play example"
```
Comment (Andersson007) [inline plugins/callback/diy.py:669]:
```suggestion
  - name: "Default plugin output: play example"
```
https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#examples-block

File: plugins/callback/diy.py [hunk 12]
```diff
@@ -0,0 +1,1421 @@
...
+    hosts: localhost
+    gather_facts: no
+    tasks:
+      - name:  default plugin output
+        debug:
+          msg: default plugin output
+
+  - name: override from play vars
```
Comment (Andersson007) [inline plugins/callback/diy.py:677]:
```suggestion
  - name: Override from play vars
```

File: plugins/callback/diy.py [hunk 13]
```diff
@@ -0,0 +1,1421 @@
...
+  # Newline after every callback
+  # on_any_msg='{{ " " | join("\n") }}'
+
+playbook.yml: >
+  ---
+  - name: "default plugin output: play example"
+    hosts: localhost
+    gather_facts: no
+    tasks:
+      - name:  default plugin output
```
Comment (Andersson007) [inline plugins/callback/diy.py:673]:
```suggestion
      - name:  Default plugin output
```

File: plugins/callback/diy.py [hunk 14]
```diff
@@ -0,0 +1,1421 @@
...
+        {% set prefix = "" %}
+        {% set suffix = "" %}
+        {% set color_two = "" %}
+        {% endif %}
+        {{ color_one }}{{ "%s%s%s" | format(prefix,key,suffix) }}{{ color_two }}: {{ ansible_callback_diy.stats[key] | to_nice_yaml }}
+        {% endif %}
+        {% endfor %}
+
+    tasks:
+      - name: custom banner with default plugin result output
```
Comment (Andersson007) [inline plugins/callback/diy.py:730]:
```suggestion
      - name: Custom banner with default plugin result output
```

File: plugins/callback/diy.py [hunk 15]
```diff
@@ -0,0 +1,1421 @@
...
+        debug:
+          msg: "default plugin output: result example"
+
+      - name: override from task vars
```
Comment (Andersson007) [inline plugins/callback/diy.py:734]:
```suggestion
      - name: Override from task vars
```

File: plugins/callback/diy.py [hunk 16]
```diff
@@ -0,0 +1,1421 @@
...
+        changed_when: true
+        vars:
+          white_fg_red_bg: "\e[0m\e[48;5;1m"
+          two: "{{ white_fg_red_bg }}    2    "
+          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
+          ansible_callback_diy_playbook_on_task_start_msg_color: bright magenta
+          ansible_callback_diy_runner_on_ok_msg: "diy output(via task vars): result example: \n{{ ansible_callback_diy.result.output.msg }}\n"
+          ansible_callback_diy_runner_on_ok_msg_color: "{{ 'yellow' if ansible_callback_diy.result.is_changed else 'bright green' }}"
+
+      - name: suppress output
```
Comment (Andersson007) [inline plugins/callback/diy.py:746]:
```suggestion
      - name: Suppress output
```

File: plugins/callback/diy.py [hunk 17]
```diff
@@ -0,0 +1,1421 @@
...
+        debug:
+          msg: i should not be displayed
+        vars:
+          ansible_callback_diy_playbook_on_task_start_msg: ""
+          ansible_callback_diy_runner_on_ok_msg: ""
+
+      - name: using alias vars (see ansible.cfg)
```
Comment (Andersson007) [inline plugins/callback/diy.py:753]:
```suggestion
      - name: Using alias vars (see ansible.cfg)
```

File: plugins/callback/diy.py [hunk 18]
```diff
@@ -0,0 +1,1421 @@
...
+        debug:
+          msg:
+        when: False
+        vars:
+          ansible_callback_diy_playbook_on_task_start_msg: ""
+          on_skipped_msg: "diy output(via task vars): skipped example:\n\e[0m\e[38;5;4m\u25b6\u25b6 {{ ansible_callback_diy.result.task.name }}\n"
+          on_skipped_msg_color: white
+
+      - name: just stdout
```
Comment (Andersson007) [inline plugins/callback/diy.py:762]:
```suggestion
      - name: Just stdout
```

File: plugins/callback/diy.py [hunk 19]
```diff
@@ -0,0 +1,1421 @@
...
+          ansible_callback_diy_playbook_on_task_start_msg: "\ndiy output(via task vars): task example: {{ ansible_callback_diy.task.name }}"
+          multiline: "line\nline\nline"
+          ansible_callback_diy_runner_on_ok_msg: |+2
+            some
+            {{ ansible_callback_diy.result.output.msg }}
+            output
+
+          ansible_callback_diy_playbook_on_task_start_msg_color: bright blue
+
+      - name: indentation
```
Comment (Andersson007) [inline plugins/callback/diy.py:781]:
```suggestion
      - name: Indentation
```
and so on for all `-name: Must start with a capital letter`

File: plugins/callback/diy.py [hunk 20]
```diff
@@ -0,0 +1,1421 @@
...
+        )
+
+        if self._using_diy(spec=self._diy_spec):
+            self._output(spec=self._diy_spec)
+
+        if self._parent_has_callback():
+            with self._suppress_stdout(enabled=self._using_diy(spec=self._diy_spec)):
+                super(CallbackModule, self).v2_playbook_on_handler_task_start(task)
+
+    def v2_playbook_on_vars_prompt(self, varname, private=True, prompt=None, encrypt=None, confirm=False, salt_size=None, salt=None, default=None, unsafe=None):
```
Comment (Andersson007) [inline plugins/callback/diy.py:1327]:
```suggestion
    def v2_playbook_on_vars_prompt(self, varname, private=True, prompt=None, encrypt=None,
                                   confirm=False, salt_size=None, salt=None, default=None, unsafe=None):
```
Let's make all such long lines a bit shorter where possible?

**Comment by Andersson007** [comment]
@theque5t thanks for the PR!
If this is the same, could you please close https://github.com/ansible/ansible/pull/65844 ?

**Comment by Andersson007** [comment]
@felixfontein what do you think whether it can be unit tested?

**Comment by felixfontein** [comment]
@Andersson007 I think integration tests would be better! I think the `default` callback tests (https://github.com/ansible/ansible/tree/devel/test/integration/targets/callback_default) are a good pointer to how callbacks can be tested.

**Comment by Andersson007** [comment]
> @Andersson007 I think integration tests would be better! I think the `default` callback tests (https://github.com/ansible/ansible/tree/devel/test/integration/targets/callback_default) are a good pointer to how callbacks can be tested.

@felixfontein thanks for this! @theque5t ^

**Comment by theque5t** [comment]
> > @Andersson007 I think integration tests would be better! I think the `default` callback tests (https://github.com/ansible/ansible/tree/devel/test/integration/targets/callback_default) are a good pointer to how callbacks can be tested.
> 
> @felixfontein thanks for this! @theque5t ^

Thanks for all the feedback @Andersson007 and @felixfontein! I'll work on applying the changes and follow-up.

**Comment by theque5t** [comment]
Cc: @felixfontein @Andersson007 

Review 1 Todos:
- [ ] Remove `ANSIBLE_METADATA` block
- [ ] The plugin `description`  should end in a trailing dot, and the wording should be adjusted to exclude `for the DIYer, because it`
- [ ] Change references of `L(default,./default.html)` to `C(default)`  in the plugin `notes` section
- [ ] Change references of `C(ansible_callback_diy)` to `I(ansible_callback_diy)`
- [ ] Refactor the section that describes the option behavior using multiple lines, and add commas to if/then statements like so `if ..., then ...`
- [ ] Change reference of `./default.html` to https://docs.ansible.com/ansible/latest/plugins/callback/default.html in the plugin `seealso` section
- [ ] All option descriptions should be written in full sentences with proper capitalization and trailing dot
- [ ] Remove all references of `required: False` on options, because `required: no` is the default
- [ ] Change all references of `type: string`  on options to `type: str`
- [ ] Change all references of option names in option descriptions to use `I()` formatting
- [ ] Change reference of `stdout_callback=diy` in ansible.cfg example to `stdout_callback=community.general.diy`
- [ ] Each example task/play should include a name line
- [ ] In the play and task names in the example, the `name:` line should be capitalized and not include a trailing dot.
- [ ] Fix misspelled `multline` in example
- [ ] Ensure the multiline string beneath `class CallbackModule(Default)` matches the plugin documentation description
- [ ] Change the longer lines to be shorter where possible
- [ ] Close ansible/ansible#65844
- [ ] Add integration tests using the `default` callback tests as an example
- [ ] Try to move `class DummyStdout(object)`  logic intent 1 level up

File: plugins/callback/diy.py (hunk 2 above)
Comment (theque5t) [reply plugins/callback/diy.py]: @felixfontein My apologies if I'm misunderstanding something, but technically `ansible_callback_diy` is not an option name. It is a dictionary name that can be used as part of a value that would be supplied to an option. Therefore, I think it should stay typeset with `C(...)`? Or, perhaps I'm not understanding the scope of an "option name"?

File: plugins/callback/diy.py (hunk 5 above)
Comment (theque5t) [reply plugins/callback/diy.py:808]: @felixfontein Just want to get clarification on your suggestion here. I'm not sure if you're suggesting to just move the `DummyStdout` class up to the base class? Or if you're suggesting to move the `DummyStdout` logic's intent up to the base class (e.g. add a control that would prevent output)? Or perhaps something else entirely, and I'm completely misunderstanding 😄

**Comment by felixfontein** [comment]
Please check the failing sanity tests (https://app.shippable.com/github/ansible-collections/community.general/runs/1831/1/tests and https://app.shippable.com/github/ansible-collections/community.general/runs/1831/4/tests).

**Comment by theque5t** [comment]
rebuild

**Comment by theque5t** [comment]
> Cc: @felixfontein @Andersson007
> 
> Review 1 Todos:
> 
> * [x]  Remove `ANSIBLE_METADATA` block
> * [x]  The plugin `description`  should end in a trailing dot, and the wording should be adjusted to exclude `for the DIYer, because it`
> * [x]  Change references of `L(default,./default.html)` to `C(default)`  in the plugin `notes` section
> * [x]  Refactor the section that describes the option behavior using multiple lines, and add commas to if/then statements like so `if ..., then ...`
> * [x]  Change reference of `./default.html` to https://docs.ansible.com/ansible/latest/plugins/callback/default.html in the plugin `seealso` section
> * [x]  All option descriptions should be written in full sentences with proper capitalization and trailing dot
> * [x]  Remove all references of `required: False` on options, because `required: no` is the default
> * [x]  Change all references of `type: string`  on options to `type: str`
> * [x]  Change all references of option names in option descriptions to use `I()` formatting
> * [x]  Change reference of `stdout_callback=diy` in ansible.cfg example to `stdout_callback=community.general.diy`
> * [x]  Each example task/play should include a name line
> * [x]  In the play and task names in the example, the `name:` line should be capitalized and not include a trailing dot.
> * [x]  Fix misspelled `multline` in example
> * [x]  Ensure the multiline string beneath `class CallbackModule(Default)` matches the plugin documentation description
> * [x]  Change `CALLBACK_NAME` value from `diy` to `community.general.diy`
> * [x]  Close [ansible/ansible#65844](https://github.com/ansible/ansible/pull/65844)
> * [x]  Change the longer lines to be shorter where possible
> * [x]  Get sanity tests passing
> * [x]  Add integration tests using the `default` callback tests as an example
> * Pending Clarification: Change references of `C(ansible_callback_diy)` to `I(ansible_callback_diy)`
> * Pending Clarification: Try to move `class DummyStdout(object)`  logic intent 1 level up

@felixfontein @Andersson007 ^ I believe I've applied all the necessary changes with the exception of the 2 "Pending Clarification" points.

**Comment by theque5t** [comment]
ready_for_review

File: plugins/callback/diy.py (hunk 5 above)
Comment (felixfontein) [reply plugins/callback/diy.py:808]: No, put `DummyStdout` to the top level of this file. Inner classes are not something that is common in Python, and so this is confusing more people looking at your code than helping.

File: plugins/callback/diy.py (hunk 2 above)
Comment (felixfontein) [reply plugins/callback/diy.py]: Ok, I think now I understand. Use `C(...)` in that case.

Comment (theque5t) [reply plugins/callback/diy.py]: Excellent, will do.

File: plugins/callback/diy.py (hunk 5 above)
Comment (theque5t) [reply plugins/callback/diy.py]: Understood. Thank you for the clarification. I made the change.

**Comment by theque5t** [comment]
> Cc: @felixfontein @Andersson007
> 
> Review 1 Todos:
> 
> * [x]  Remove `ANSIBLE_METADATA` block
> * [x]  The plugin `description`  should end in a trailing dot, and the wording should be adjusted to exclude `for the DIYer, because it`
> * [x]  Change references of `L(default,./default.html)` to `C(default)`  in the plugin `notes` section
> * [x]  Refactor the section that describes the option behavior using multiple lines, and add commas to if/then statements like so `if ..., then ...`
> * [x]  Change reference of `./default.html` to https://docs.ansible.com/ansible/latest/plugins/callback/default.html in the plugin `seealso` section
> * [x]  All option descriptions should be written in full sentences with proper capitalization and trailing dot
> * [x]  Remove all references of `required: False` on options, because `required: no` is the default
> * [x]  Change all references of `type: string`  on options to `type: str`
> * [x]  Change all references of option names in option descriptions to use `I()` formatting
> * [x]  Change reference of `stdout_callback=diy` in ansible.cfg example to `stdout_callback=community.general.diy`
> * [x]  Each example task/play should include a name line
> * [x]  In the play and task names in the example, the `name:` line should be capitalized and not include a trailing dot.
> * [x]  Fix misspelled `multline` in example
> * [x]  Ensure the multiline string beneath `class CallbackModule(Default)` matches the plugin documentation description
> * [x]  Change `CALLBACK_NAME` value from `diy` to `community.general.diy`
> * [x]  Close [ansible/ansible#65844](https://github.com/ansible/ansible/pull/65844)
> * [x]  Change the longer lines to be shorter where possible
> * [x]  Get sanity tests passing
> * [x]  Add integration tests using the `default` callback tests as an example
> * [x]  Move `DummyStdout` to the top level of the file

@felixfontein @Andersson007 ^ I believe I've applied all the necessary changes.

Re: 2 failing builds
- Build 1992.47: Related to `firewalld` integration tests I believe.
- Build 1992.61: Related to `mysql_replication` integration tests I believe.

**Comment by theque5t** [comment]
ready_for_review

**Comment by Andersson007** [comment]
i close-reopened the pr to relaunch CI tests

File: plugins/callback/diy.py [hunk 21]
```diff
@@ -0,0 +1,1420 @@
...
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+  callback: diy
+  callback_type: stdout
+  short_description: Customize the output
+  description:
+    - Callback plugin that allows you to supply your own custom callback templates to be output.
+  version_added: "2.10"
```
Comment (Andersson007) [inline plugins/callback/diy.py:15]:
```suggestion
```
there will be its own versioning in collections

**Comment by Andersson007** [comment]
@felixfontein do you have any questions/suggestions about CI tests?

**Comment by theque5t** [comment]
bot_status

**Comment by theque5t** [comment]
ready_for_review

File: plugins/callback/diy.py [hunk 22]
```diff
@@ -0,0 +1,1419 @@
...
+# Copyright: (c) 2019, Trevor Highfill <trevor.highfill@outlook.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+  callback: diy
+  callback_type: stdout
+  short_description: Customize the output
```
Comment (felixfontein) [inline plugins/callback/diy.py:12]: We plan to do a release the coming week, for that we'd need:
```suggestion
  short_description: Customize the output
  version_added: 0.2.0
```

**Comment by theque5t** [comment]
ready_for_review

**Comment by Andersson007** [comment]
@theque5t thanks for the contribution!
@felixfontein thanks for reviewing!

**Comment by Andersson007** [comment]
merged https://github.com/ansible-collections/community.general/pull/432 into master

**Comment by yzcyayaya** [comment]
How do I print status of task when using until with ansible?

**Pull request merged**